src/App.spec.ts

```diff
@@ -342,18 +344,147 @@ describe('App', () => {
         });
         app.use(fakeSecondMiddleware);
         app.error(fakeErrorHandler);
-        fakeReceiver.emit('message', dummyReceiverEvent);
-        await delay();
+        await fakeReceiver.sendEvent(dummyReceiverEvent);
 
         // Assert
         assert.instanceOf(fakeErrorHandler.firstCall.args[0], Error);
       });
+
+      it('correctly waits for async listeners', async () => {
+        let changed = false;
+
+        app.use(async ({ next }) => {
+          await delay(100);
+          changed = true;
+
+          await next();
+        });
+
+        await fakeReceiver.sendEvent(dummyReceiverEvent);
+        assert.isTrue(changed);
+        assert(fakeErrorHandler.notCalled);
+      });
+
+      it('throws errors which can be caught by upstream async listeners', async () => {
+        const thrownError = new Error('Error handling the message :(');
+        let caughtError;
+
+        app.use(async ({ next }) => {
+          try {
+            await next();
+          } catch (err) {
+            caughtError = err;
+          }
+        });
+
+        app.use(async () => {
+          throw thrownError;
+        });
+
+        app.error(fakeErrorHandler);
+
+        await fakeReceiver.sendEvent(dummyReceiverEvent);
+
+        assert.equal(caughtError, thrownError);
+        assert(fakeErrorHandler.notCalled);
+      });
+
+      it('calls async middleware in declared order', async () => {
+        const message = ':wave:';
+        let middlewareCount = 0;
+
+        const assertOrderMiddleware = (orderDown: number, orderUp: number) => async ({ next }: any) => {
+          await delay(100);
+          middlewareCount += 1;
+          assert.equal(middlewareCount, orderDown);
+          await next();
+          middlewareCount += 1;
+          assert.equal(middlewareCount, orderUp);
+        };
+
```

Comment:
i can see that the number arguments are used to declare the expected order. can we add a comment to clarify that? it took me a while to understand this.